lib/Runtime/Library/JavascriptObject.cpp

```diff
@@ -271,17 +274,20 @@ using namespace Js;
                 object->GetLibrary()->GetTypesWithOnlyWritablePropertyProtoChainCache()->Clear();
             }
 
-            if (!objectAndPrototypeChainHasOnlyWritableDataProperties)
+            if (!allProtoCachesInvalidated)
             {
                 // Invalidate StoreField/PropertyGuards for any non-WritableData property in the new chain
-                JavascriptOperators::MapObjectAndPrototypes<true>(newPrototype, [=](RecyclableObject* obj)
+                JavascriptOperators::MapObjectAndPrototypesUntil<true>(newPrototype, [&](RecyclableObject* obj)->bool
                 {
-                    if (!obj->HasOnlyWritableDataProperties())
-                    {
-                        obj->AddToPrototype(scriptContext);
-                    }
+                    obj->AddToPrototype(scriptContext, &allProtoCachesInvalidated);
+                    return allProtoCachesInvalidated;
                 });
             }
+
+            JavascriptOperators::MapObjectAndPrototypesUntil<true>(object->GetPrototype(), [=](RecyclableObject* obj)->bool
```

Comment:
nit: empty capture list could show more clearly that we don't require anything from the enclosing scope

Reply:
Done. Thanks.